src/AppContext.tsx

```diff
@@ -0,0 +1,193 @@
+
+  const clearCompleted = useCallback(
+    () => {
+      const completedTodos = todos
+        .filter(todoToFind => todoToFind.completed);
+
+      setSelectedTodoIds(currentIds => (
+        [...currentIds, ...completedTodos
+          .map(completedTodo => completedTodo.id),
+        ]
+      ));
+
+      completedTodos.map(completedTodo => todoService
+        .deletTodo(completedTodo.id));
+
+      setTimeout(() => {
+        setTodos(currentTodos => currentTodos.filter(
+          todoToFilter => !todoToFilter.completed,
+        ));
+      }, 500);
+    }, [todos],
+  );
+
+  const createNewTodo = (title: string) => {
+    setIsLoading(true);
+    setSelectedTodoIds(ids => [...ids, 0]);
+    setTempTodo({
+      id: 0,
+      userId: USER_ID,
+      title,
+      completed: false,
+    });
+
+    todoService.createTodo({
+      userId: USER_ID,
+      title,
+      completed: false,
+    })
+      .then(newTodo => {
+        setTodoTitle('');
+        setTimeout(() => {
+          setTodos(currentTodo => [...currentTodo, newTodo]);
+        }, 500);
+      })
+      .catch(() => {
+        setSelectedTodoIds(ids => ids.filter(
+          id => id !== 0,
+        ));
+        setErrorMessage(ErrorType.UnableToAddTodo);
+        setTimeout(() => setErrorMessage(null), 2000);
+      })
+      .finally(() => {
+        setIsLoading(false);
+        setSelectedTodoIds(ids => ids.filter(
+          id => id !== 0,
+        ));
+        setTimeout(() => setTempTodo(null), 500);
+      });
+  };
+
+  const deleteTodo = useCallback(
+    (todoId: number) => {
+      setSelectedTodoIds(currentIds => [...currentIds, todoId]);
+      todoService.deletTodo(todoId)
+        .then(() => {
+          setTimeout(() => {
+            setTodos(currentTodos => currentTodos
+              .filter(post => post.id !== todoId));
+          }, 500);
+        })
+        .catch(() => {
+          setSelectedTodoIds(ids => {
+            ids.splice(ids.indexOf(todoId), 1);
+
+            return ids;
+          });
+          setErrorMessage(ErrorType.UnableToDeleteTodo);
+          setTimeout(() => setErrorMessage(null), 2000);
+        })
+        .finally(() => setTimeout(
+          () => setSelectedTodoIds(ids => ids.splice(ids.indexOf(todoId), 1)),
+          500,
+        ));
+    }, [],
+  );
+
+  const value = useMemo(() => ({
+    todos,
+    setTodos,
+    filterBy,
+    setFilterBy,
+    errorMessage,
+    setErrorMessage,
+    clearCompleted,
+    setSelectedTodoIds,
+    todoTitle,
+    isLoading,
+    createNewTodo,
+    setTodoTitle,
+    tempTodo,
+    deleteTodo,
+    selectedTodoIds,
+    setTempTodo,
+
+  }), [
+    todos,
+    setTodos,
+    filterBy,
+    setFilterBy,
+    errorMessage,
+    setErrorMessage,
+    clearCompleted,
+    setSelectedTodoIds,
+    todoTitle,
+    isLoading,
+    createNewTodo,
+    setTodoTitle,
+    tempTodo,
+    deleteTodo,
+    selectedTodoIds,
+  ]);
```

Comment:
👀

Reply:
@MrHunterBoi better to leave more meaningful comments